### ISSUE TITLE
Shake instead of Make?

### ISSUE THREAD

**Issue by alanz** [haskell/haskell-ide-engine#650]
I understand installation of the various versions of hie via a Makefile is seen as being linux-centric, and causes problems on e.g. windows.

Could we avert this by using something like Shake?

And if so, is someone prepared to do it?

**Comment by Anrock** [comment]
tl;dr: What exact problems are we trying to solve here?

Most problems on Windows i've encountered were caused by either filepath handling (hie bugs) or hie not being buildable due to linker errors or deps mistmatch (stack problems).
Windows specific bugs with Makefile were trivially fixable.

Shake may tidy up or speed up or *improve* building somehow else, but i don't know what *problems* it will solve.

**Comment by alanz** [comment]
@Anrock I think the first point of the discussion is to see if it is actually a problem.

**Comment by IronGremlin** [comment]
Might I suggest just directing users to attempt to run `stack exec make build-all` instead?


This should work just fine regardless of platform, and would start generating error reports from windows users if it doesn't work for them.

It's a bit of a [purple beggar](http://thecodelesscode.com/case/170) strategy, but you're really not doing anything terribly complicated in that makefile right now, and it really should work.... Better to fail early and loudly than to pass along silently.

**Comment by ShrykeWindgrace** [comment]
@Anrock In my opinion, the question is whether we want to force `make` on windows users. It is not installed by default (and you have several versions of `make` for windows that clash with each other), so Occam's razor suggests that a `shake` solution is better.

Then again, some ports of `make` on windows use `sh` implicitly, which (a) not guaranteed to exist on the system (b) not guaranteed to behave exactly as `make` expects (fell in this trap a couple of days ago). The issue can be solved by launching `make` from e.g. `git-bash`, which is yet another redundant tool that we need to install. I skimmed  over the makefile script, I did not see any magic that could not be solved with haskell-based tools.

TLDR: to use `make` reliably on windows, users need to install a lot of stuff that will not be used later at all. `Shake` is a purely haskell solution, and we know for sure that by this point haskell is installed on the system.

I'll try to sketch a `shake` build script by the end of August, no promises to do it earlier, summer is upon us!=)

**Comment by ShrykeWindgrace** [comment]
@IronGremlin 
> `stack exec make build-all`... This should work just fine regardless of platform

Not  really, `make` is not directly available on windows, you need to install it manually.

**Comment by Anrock** [comment]
@ShrykeWindgrace yup, make on Windows is one of valid points, see discussion #669. Though it's not so bad actually - stack brings msys and installing `make` is only one `pacman -S make` away, so user doesn't have to "install a lot of stuff that will not be used later".

**Comment by ShrykeWindgrace** [comment]
@Anrock I agree that we can rely on `stack exec -- pacman -S make` to get `make`; I do not quite agree that we need `make` at all (even for unix).

To contradict a bit my previous comment: I doubt that the framework for building `hie` would evolve frequently, hence once a user runs all magic commands to set up `hie` compilation environment (imagine a user that never heard of `pacman` or `make`!), they are able to continue compiling it, so the long-term gain of `shake` is unclear for me.

**Comment by fendor** [comment]
I implemented a proof of concept Shakefile in my fork of haskell-ide-engine. It is still WIP, and I have no access to a Mac Machine.
What currently works is building the docs, building different versions of hie, as well as building everything at once. 
What's currently debatable is the implementation of the help message in the Shakefile.
Also. how to actually invoke the `Shakefile.hs` is up for discussion. 
Possibilities are: 
* It can be a subpackage that is built with stack
* It can be executed via `runhaskell` relying on the user to supply a fitting environment.
* It can be a stack script (`#!/usr/bin/env stack`)

**Comment by fendor** [comment]
@lorenzo Do you know what the icu related targets in the Makefiles are?

**Comment by fendor** [comment]
I implemented the Shakefile to be run directly from `stack`.

```
# one-stop-shop build everything
stack Shakefile.hs build-all
```

Features:

* `stack` is the only build requirement
* `shake` is used instead of `make`
* should work on all platforms (not fully tested yet)
* submodules are synced / updated at the start -> works after clean clone of project

Quirks:

* `shake.yaml` is necessary to start the `stack Shakefile.hs ...` command before submodules are loaded.
